crates/matrix-sdk/src/event_cache/paginator.rs

```diff
@@ -0,0 +1,564 @@
+    static USER_ID: Lazy<&UserId> = Lazy::new(|| user_id!("@paul:atreid.es"));
+
+    #[async_trait]
+    impl PaginableRoom for DummyRoom {
+        async fn event_with_context(
+            &self,
+            event_id: &EventId,
+            _lazy_load_members: bool,
+        ) -> Result<EventWithContextResponse, PaginatorError> {
+            // Wait for the room to be marked as ready first.
+            self.room_ready.notified().await;
+
+            let event_factory = EventFactory::new().room(*ROOM_ID).sender(*USER_ID);
+
+            let event = event_factory.text_msg("hello!").event_id(event_id).into_raw_timeline();
+
+            let before = (0..10)
+                .rev()
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
+                })
+                .collect();
+
+            let after = (10..20)
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
+                })
+                .collect();
+
+            return Ok(EventWithContextResponse {
+                event: Some(TimelineEvent::new(event)),
+                events_before: before,
+                events_after: after,
+                prev_batch_token: Some("prev".to_owned()),
+                next_batch_token: Some("next".to_owned()),
+                state: Vec::new(),
+            });
+        }
+
+        async fn messages(&self, opts: MessagesOptions) -> Result<Messages, PaginatorError> {
+            assert_eq!(
+                opts.from.as_deref(),
+                Some("prev"),
+                "we must receive a single back-pagination"
+            );
+            assert_eq!(opts.dir, Direction::Backward, "we must receive a single back-pagination");
+
+            self.room_ready.notified().await;
+
+            let event_factory = EventFactory::new().room(*ROOM_ID).sender(*USER_ID);
+
+            let chunk = (20..30)
+                .rev()
+                .map(|i| {
+                    TimelineEvent::new(event_factory.text_msg(format!("{i}")).into_raw_timeline())
+                })
+                .collect();
+
+            return Ok(Messages {
+                start: opts.from.unwrap().to_owned(),
+                end: None,
+                chunk,
+                state: Vec::new(),
+            });
+        }
+    }
+
+    async fn assert_invalid_state<T: std::fmt::Debug>(
+        task: impl Future<Output = Result<T, PaginatorError>>,
+        expected: PaginatorState,
+        actual: PaginatorState,
+    ) {
+        assert_let!(
+            Err(PaginatorError::InvalidPreviousState {
+                expected: real_expected,
+                actual: real_actual
+            }) = task.await
+        );
+        assert_eq!(real_expected, expected);
+        assert_eq!(real_actual, actual);
+    }
+
+    #[async_test]
+    async fn test_state() {
```

Comment:
Please can we have more tests for paginating multiple times (successfully), hitting the end, paginating forwards, and paginating in both directions?